core/templates/services/playthrough.service.spec.ts

```diff
@@ -16,236 +16,463 @@
-        playthroughService.recordAnswerSubmitAction(
-          'stateName1', 'stateName1', 'TextInput', 'Hello', 'Try again', 30);
-        playthroughService.recordAnswerSubmitAction(
-          'stateName1', 'stateName1', 'TextInput', 'Hello', 'Try again', 30);
-        playthroughService.recordAnswerSubmitAction(
-          'stateName1', 'stateName1', 'TextInput', 'Hello', 'Try again', 30);
-        playthroughService.recordAnswerSubmitAction(
-          'stateName1', 'stateName1', 'TextInput', 'Hello', 'Try again', 30);
+    describe('Issue identification', () => {
+      it('should return null issue if playthrough has no problems', () => {
+        this.mockExplorationTimer(400);
+        playthroughService.recordExplorationStartAction('A');
+        this.recordStateTransitions('ABC');
+        playthroughService.recordExplorationQuitAction('C', 60);
+
+        let playthrough = playthroughService.getPlaythrough();
+        expect(playthrough).not.toBeNull();
+        expect(playthrough.issueType).toBeNull();
+        expect(playthrough.issueCustomizationArgs).toBeNull();
+      });
 
-        playthroughService.recordPlaythrough(false);
+      it('should identify multiple incorrect submissions', () => {
+        this.mockExplorationTimer(400);
+        playthroughService.recordExplorationStartAction('A');
+        this.recordIncorrectAnswers('A', 5);
+        playthroughService.recordExplorationQuitAction('A', 60);
 
         let playthrough = playthroughService.getPlaythrough();
         expect(playthrough.issueType).toEqual('MultipleIncorrectSubmissions');
         expect(playthrough.issueCustomizationArgs).toEqual({
-          state_name: {value: 'stateName1'},
-          num_times_answered_incorrectly: {value: 5}
+          state_name: {value: 'A'},
+          num_times_answered_incorrectly: {value: 5},
         });
       });
 
-      it('identifies early quits', () => {
-        playthroughService.recordExplorationStartAction('stateName1');
-        playthroughService.recordAnswerSubmitAction(
-          'stateName1', 'stateName1', 'TextInput', 'Hello', 'Try again', 30);
-        playthroughService.recordExplorationQuitAction('stateName1', 60);
+      it('should return null if state with multiple incorrect submissions is ' +
+        'eventually completed', () => {
+        this.mockExplorationTimer(400);
+        playthroughService.recordExplorationStartAction('A');
+        this.recordIncorrectAnswers('A', 5);
+        this.recordStateTransitions('ABC');
+        playthroughService.recordExplorationQuitAction('C', 60);
+
+        let playthrough = playthroughService.getPlaythrough();
+        expect(playthrough.issueType).toBeNull();
+        expect(playthrough.issueCustomizationArgs).toBeNull();
+      });
+
+      it('should identify cyclic state transitions', () => {
+        this.mockExplorationTimer(400);
+        playthroughService.recordExplorationStartAction('A');
+        this.recordCycle('ABC', 3);
```

Comment:
I think you should represent 'ABC' as ['A', 'B', 'C']. It's more semantically correct. There is no requirement that state names are 1-letter long, so this is confusing as written.

Ditto elsewhere.